cmd/promtool/backfill.go

```
@@ -0,0 +1,162 @@
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/tsdb"
+)
+
+// Parser is the openmetrics parser.
+type Parser struct {
+	textparse.Parser
+	s *bufio.Scanner
+}
+
+// NewParser is a tiny layer between textparse.Parser and importer.Parser which works on io.Reader.
+func NewParser(r io.Reader) textparse.Parser {
+	return &Parser{s: bufio.NewScanner(r)}
+}
+
+// Next advances the parser to the next sample. It returns io.EOF if no
+// more samples were read.
+func (p *Parser) Next() (textparse.Entry, error) {
+	for p.s.Scan() {
+		line := p.s.Bytes()
+		line = append(line, '\n')
+		p.Parser = textparse.New(line, string(expfmt.FmtOpenMetrics))
+		if et, err := p.Parser.Next(); err != io.EOF {
+			return et, err
+		}
+	}
+	if err := p.s.Err(); err != nil {
+		return 0, err
+	}
+	return 0, io.EOF
+}
+
+func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
+	var maxt, mint int64 = math.MinInt64, math.MaxInt64
+	for {
+		entry, err := p.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, 0, errors.Wrap(err, "next")
+		}
+		if entry != textparse.EntrySeries {
+			continue
+		}
+		l := labels.Labels{}
+		p.Metric(&l)
+		_, ts, _ := p.Series()
+		if ts == nil {
+			return 0, 0, errors.Errorf("expected timestamp for series %v, got none", l.String())
+		}
+		if *ts > maxt {
+			maxt = *ts
+		}
+		if *ts < mint {
+			mint = *ts
+		}
+	}
+	if maxt == math.MinInt64 {
+		maxt = 0
+	}
+	if mint == math.MaxInt64 {
+		mint = 0
+	}
+	return maxt, mint, nil
+}
+
+func createBlocks(input *os.File, mint, maxt int64, outputDir string) error {
+	var offset int64 = 2 * time.Hour.Milliseconds()
+	mint = offset * (mint / offset)
+	for t := mint; t <= maxt; t = t + offset {
+		err := func() error {
+			w, errw := tsdb.NewBlockWriter(log.NewNopLogger(), outputDir, offset)
+			if errw != nil {
+				return errors.Wrap(errw, "block writer")
+			}
+			defer w.Close()
```

Let's not ignore the error here. You could have a named error `func() (err error)` and use `tsdb_errors.NewMulti` to combine the errors. Example:
```go
defer func(){
    err = tsdb_errors.NewMulti(err, w.Close()).Err()
}()
```